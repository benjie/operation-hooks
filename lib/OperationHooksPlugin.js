"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const OperationHooksCorePlugin_1 = require("./OperationHooksCorePlugin");
const OperationMessagesPlugin_1 = require("./OperationMessagesPlugin");
const OperationMessagesMutationPayloadPlugin_1 = require("./OperationMessagesMutationPayloadPlugin");
const OperationMessagesMutationPreFlightPlugin_1 = require("./OperationMessagesMutationPreFlightPlugin");
const PgOperationHooksPlugin_1 = require("./PgOperationHooksPlugin");
const graphile_utils_1 = require("graphile-utils");
// Export types
__export(require("./OperationHooksCorePlugin"));
const OperationHooksPlugin = graphile_utils_1.makePluginByCombiningPlugins(OperationHooksCorePlugin_1.default, OperationMessagesPlugin_1.default, PgOperationHooksPlugin_1.default, OperationMessagesMutationPayloadPlugin_1.default, OperationMessagesMutationPreFlightPlugin_1.default);
OperationHooksPlugin.displayName = "OperationHooksPlugin";
exports.default = OperationHooksPlugin;
