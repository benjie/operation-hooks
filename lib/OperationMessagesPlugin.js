"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const addMessagesToMeta = (_, _args, _context, resolveInfo) => {
    resolveInfo.graphileMeta.messages = [];
    return _;
};
const validateMessages = (_, _args, _context, resolveInfo) => {
    const firstError = resolveInfo.graphileMeta.messages.find(message => message.level === "error");
    if (firstError) {
        throw new Error(`Aborting ${resolveInfo.fieldName} due to error: ${firstError.message}`);
    }
    return _;
};
const addMessagesToError = (error, _args, _context, resolveInfo) => {
    error.extensions = error.extensions || {};
    error.extensions.messages = resolveInfo.graphileMeta.messages;
    return error;
};
const MutationMessagesPlugin = function MutationMessagesPlugin(builder) {
    builder.hook("init", (_, build) => {
        build.addOperationHook(() => {
            return {
                before: [
                    { priority: 100, callback: addMessagesToMeta },
                    { priority: 900, callback: validateMessages },
                ],
                after: [{ priority: 900, callback: validateMessages }],
                error: [{ priority: 500, callback: addMessagesToError }],
            };
        });
        return _;
    });
};
exports.default = MutationMessagesPlugin;
