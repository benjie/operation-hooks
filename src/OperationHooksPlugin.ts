import { Plugin, Context } from "graphile-build";
import {
  GraphQLResolveInfo,
  GraphQLFieldResolver,
  GraphQLObjectType,
} from "graphql";

export interface GraphQLResolveInfoWithMeta extends GraphQLResolveInfo {
  graphileMeta: {};
}

export type OperationHookCallback<
  T = any,
  TArgs = { [argName: string]: any }
> = (
  input: T,
  args: TArgs,
  context: any,
  resolveInfo: GraphQLResolveInfoWithMeta
) => T | null | Promise<T | null>;

export interface OperationHookEntry<T = any> {
  priority: number;
  callback: OperationHookCallback<T>;
}

export interface OperationHook {
  before?: Array<OperationHookEntry>;
  after?: Array<OperationHookEntry>;
  error?: Array<OperationHookEntry<Error>>;
}

export type OperationHookGenerator = (
  fieldContext: Context<any>
) => OperationHook;

// Hooks are applied one after the other, in an asynchronous chain.
async function applyHooks<T, TArgs>(
  hooks: OperationHookCallback[],
  input: T,
  args: TArgs,
  context: any,
  resolveInfo: GraphQLResolveInfoWithMeta
): Promise<T | null> {
  let output: T | null = input;
  for (const hook of hooks) {
    output = await hook(output, args, context, resolveInfo);
    if (output === undefined) {
      throw new Error("Logic error: operation hook returned 'undefined'.");
    }

    // Nulls return early
    if (output === null) {
      return null;
    }
  }
  return output;
}

function hookSort(a: OperationHookEntry, b: OperationHookEntry): number {
  return a.priority - b.priority;
}

const OperationHooksPlugin: Plugin = function OperationHooksPlugin(builder) {
  builder.hook("build", build => {
    const _operationHookGenerators: OperationHookGenerator[] = [];
    let locked = false;
    return build.extend(build, {
      addOperationHook(fn: OperationHookGenerator) {
        if (locked) {
          throw new Error(
            "Attempted to register operation hook after a hook was applied; this indicates an issue with the ordering of your plugins. Ensure that the OperationHooksPlugin and anything that depends on it come at the end of the plugins list."
          );
        }
        _operationHookGenerators.push(fn);
      },

      _getOperationHookCallbacksForContext(
        context: Context<any>
      ): null | {
        before: OperationHookCallback[];
        after: OperationHookCallback[];
        error: OperationHookCallback<Error>[];
      } {
        // Don't allow any more hooks to be registered now that one is being applied.
        locked = true;

        // Generate the hooks, and aggregate into before/after/error arrays
        const generatedHooks: OperationHook[] = _operationHookGenerators
          .map(gen => gen(context))
          .filter(_ => _);
        const before: OperationHookEntry[] = [];
        const after: OperationHookEntry[] = [];
        const error: OperationHookEntry<Error>[] = [];
        generatedHooks.forEach(oneHook => {
          if (oneHook.before) {
            before.push(...oneHook.before);
          }
          if (oneHook.after) {
            after.push(...oneHook.after);
          }
          if (oneHook.error) {
            error.push(...oneHook.error);
          }
        });

        // No relevant hooks, don't bother wrapping the resolver
        if (before.length === 0 && after.length === 0 && error.length === 0) {
          return null;
        }

        // Sort the hooks based on their priority (remember sort() mutates the arrays)
        before.sort(hookSort);
        after.sort(hookSort);
        error.sort(hookSort);

        // Return the relevant callbacks
        return {
          before: before.map(hook => hook.callback),
          after: after.map(hook => hook.callback),
          error: error.map(hook => hook.callback),
        };
      },
    });
  });

  builder.hook("GraphQLObjectType:fields:field", (field, build, context) => {
    const { _getOperationHookCallbacksForContext } = build;
    const {
      Self,
      scope: { fieldName, isRootQuery, isRootMutation, isRootSubscription },
    } = context;

    // We only care about root fields
    if (!isRootQuery && !isRootMutation && !isRootSubscription) {
      return field;
    }

    // Get the hook for this context
    const callbacks = _getOperationHookCallbacksForContext(context);
    if (!callbacks) {
      return field;
    }

    // Get the old resolver for us to wrap
    const oldResolve = field.resolve;
    if (!oldResolve) {
      throw new Error(
        `Default resolver found for field ${
          Self.name
        }.${fieldName}; default resolvers at the root level are not supported by operation-hooks`
      );
    }

    const resolve: GraphQLFieldResolver<any, any> = async function(
      op,
      args,
      context,
      resolveInfo: GraphQLResolveInfoWithMeta
    ) {
      // Mutating for performance reasons
      resolveInfo.graphileMeta = {};

      try {
        const symbol = Symbol("before");
        // Perform the 'before' hooks
        const beforeResult = await applyHooks(
          callbacks.before,
          symbol,
          args,
          context,
          resolveInfo
        );

        // Exit early if someone changed the result
        if (beforeResult !== symbol) {
          return beforeResult;
        }

        // Call the old resolver
        const result = await oldResolve(op, args, context, resolveInfo);

        // Perform the 'after' hooks
        const afterResult = await applyHooks(
          callbacks.after,
          result,
          args,
          context,
          resolveInfo
        );
        return afterResult;
      } catch (error) {
        // An error occured, call the 'error' hooks
        const errorResult = await applyHooks(
          callbacks.error,
          error,
          args,
          context,
          resolveInfo
        );
        throw errorResult;
      }
    };
    resolve["__asyncHooks"] = true;

    // Finally override the resolve method
    return {
      ...field,
      resolve,
    };
  });

  // Ensure all the resolvers have been wrapped (i.e. sanity check)
  builder.hook("finalize", schema => {
    const missingHooks: String[] = [];
    const types = [
      schema.getQueryType(),
      schema.getMutationType(),
      schema.getSubscriptionType(),
    ].filter(_ => _);
    types.forEach((type: GraphQLObjectType) => {
      const fields = type.getFields();
      for (const field of Object.values(fields)) {
        const { resolve } = field;
        if (!resolve || !resolve["__asyncHooks"]) {
          missingHooks.push(`${type.name}.${field.name}`);
        }
      }
    });

    if (missingHooks.length) {
      throw new Error(
        `Schema validation error: operation hooks were not added to the following fields: ${missingHooks.join(
          ", "
        )}`
      );
    }

    return schema;
  });
};

export default OperationHooksPlugin;
