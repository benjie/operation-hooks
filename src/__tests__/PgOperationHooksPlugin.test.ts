import { graphql, GraphQLSchema } from "graphql";
import { Plugin } from "graphile-build";
import {
  createPostGraphileSchema,
  withPostGraphileContext,
} from "postgraphile";
import { GraphQLResolveInfoWithMessages } from "../OperationMessagesPlugin";
import OperationHooksPlugin from "../OperationHooksPlugin";
import { makeHookPlugin } from "./common";
import { Pool } from "pg";

if (!process.env.TEST_DATABASE_URL) {
  throw new Error("TEST_DATABASE_URL envvar must be set");
}

function snapshotSanitise(o: any): any {
  if (Array.isArray(o)) {
    return o.map(snapshotSanitise);
  } else if (o && typeof o === "object") {
    let result = {};
    for (const key in o) {
      const val = o[key];
      if (key === "id" && typeof val === "number") {
        result[key] = "[NUMBER]";
      } else if (key === "nodeId" && typeof val === "string") {
        result[key] = "[NodeId]";
      } else {
        result[key] = snapshotSanitise(val);
      }
    }
    return result;
  } else {
    return o;
  }
}

let pgPool: Pool;
function sqlSearchPath(sql: string) {
  return `
  begin;
  set local search_path to operation_hooks;
  ${sql};
  commit;
  `;
}

beforeAll(async () => {
  pgPool = new Pool({
    connectionString: process.env.TEST_DATABASE_URL,
  });
  await pgPool.query(
    sqlSearchPath(`
  drop schema if exists operation_hooks cascade;
  create schema operation_hooks;

  create table users (id serial primary key, name text not null);

  create type mutation_message as (
    level text,
    message text,
    path text[],
    code text
  );

  `)
  );
});

afterAll(() => {
  pgPool.end();
});

function postgraphql(
  schema: GraphQLSchema,
  query: string,
  rootValue?: any,
  context?: any,
  variables?: {
    [key: string]: any;
  } | null,
  operationName?: string
) {
  return withPostGraphileContext(
    // @ts-ignore See postgraphile#931
    {
      pgPool,
    },
    (postgraphileContext: {}) =>
      graphql(
        schema,
        query,
        rootValue,
        {
          ...postgraphileContext,
          context,
        },
        variables,
        operationName
      )
  );
}

const getPostGraphileSchema = (plugins: Plugin[] = []) =>
  createPostGraphileSchema(pgPool, "operation_hooks", {
    appendPlugins: [OperationHooksPlugin, ...plugins],
  });

const equivalentFunctions = [
  `
  create function "mutation_createUser_before"(args json) returns setof mutation_message as $$
    select row(
      'info',
      'Pre createUser mutation; name: ' || (args -> 'input' -> 'user' ->> 'name'),
      ARRAY['input', 'name'],
      'INFO1'
    )::mutation_message;
  $$ language sql volatile set search_path from current;
  `,
  `
  create function "mutation_createUser_before"(args json) returns mutation_message[] as $$
    select ARRAY[row(
      'info',
      'Pre createUser mutation; name: ' || (args -> 'input' -> 'user' ->> 'name'),
      ARRAY['input', 'name'],
      'INFO1'
    )::mutation_message]
  $$ language sql volatile set search_path from current;
  `,
  `
  create function "mutation_createUser_before"(args json) returns table(
    level text,
    message text,
    path text[],
    code text
  ) as $$
    select 
      'info',
      'Pre createUser mutation; name: ' || (args -> 'input' -> 'user' ->> 'name'),
      ARRAY['input', 'name'],
      'INFO1';
  $$ language sql volatile set search_path from current;
  `,
  `
  create function "mutation_createUser_before"(
    in args json,
    out level text,
    out message text,
    out path text[],
    out code text
  ) as $$
    select 
      'info',
      'Pre createUser mutation; name: ' || (args -> 'input' -> 'user' ->> 'name'),
      ARRAY['input', 'name'],
      'INFO1';
  $$ language sql volatile set search_path from current;
  `,
];

describe("equivalent functions", () => {
  equivalentFunctions.forEach((sqlDef, i) => {
    const matches = sqlDef.match(
      /create function "([^"]+)"\(([^)]+)\) (?:returns ([\s\S]*) )?as \$\$/
    );
    if (!matches) throw new Error(`Don't understand SQL definition ${i}!`);
    const [, funcName, funcArgs, funcReturns] = matches;
    const dropSql = `drop function "${funcName}"(${funcArgs});`;
    const omitSql = `comment on function "mutation_createUser_before"(args json) is E'@omit';`;
    describe(`hook returning '${
      funcReturns ? funcReturns.replace(/\s+/g, " ") : "record"
    }'`, () => {
      beforeAll(() => pgPool.query(sqlSearchPath(`${sqlDef};${omitSql};`)));
      afterAll(() => pgPool.query(sqlSearchPath(dropSql)));

      test("creates messages on meta", async () => {
        let resolveInfos: GraphQLResolveInfoWithMessages[] = [];
        const schema = await getPostGraphileSchema([
          makeHookPlugin(
            (
              input,
              _args,
              _context,
              resolveInfo: GraphQLResolveInfoWithMessages
            ) => {
              resolveInfos.push(resolveInfo);
              return input;
            },
            "after",
            999
          ),
        ]);
        expect(resolveInfos.length).toEqual(0);
        const data = await postgraphql(
          schema,
          `
      mutation {
        createUser(input: { user: { name: "Bobby Tables" } }) {
          user {
            nodeId
            id
            name
          }
          messages {
            level
            message
            path
          }
        }
      }
    `
        );
        expect(data.errors).toBeFalsy();
        expect(resolveInfos.length).toEqual(1);
        // TODO: I'm committing the below INVALID SNAPSHOTS but I will fix them later!
        expect(resolveInfos[0].graphileMeta.messages).toMatchInlineSnapshot(`
Array [
  Object {
    "code": "INFO1",
    "level": "info",
    "message": "Pre createUser mutation; name: Bobby Tables",
    "path": Array [
      "input",
      "name",
    ],
  },
]
`);
        expect(snapshotSanitise(data)).toMatchInlineSnapshot(`
Object {
  "data": Object {
    "createUser": Object {
      "messages": Array [
        Object {
          "level": "info",
          "message": "Pre createUser mutation; name: Bobby Tables",
          "path": Array [
            "input",
            "name",
          ],
        },
      ],
      "user": Object {
        "id": "[NUMBER]",
        "name": "Bobby Tables",
        "nodeId": "[NodeId]",
      },
    },
  },
}
`);
      });
    });
  });
});
