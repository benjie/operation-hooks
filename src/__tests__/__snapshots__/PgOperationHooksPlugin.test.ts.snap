// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updating change name is passed correct arguments 1`] = `
Object {
  "data": Object {
    "updateUserById": Object {
      "messages": Array [
        Object {
          "level": "info",
          "message": "Pre user update mutation; old name: Uzr Vun, user request: Zeb Zob",
          "path": Array [
            "input",
            "userPatch",
            "name",
          ],
        },
        Object {
          "level": "info",
          "message": "Post user update mutation; new name: Zeb Zob, user request: Zeb Zob",
          "path": Array [
            "input",
            "userPatch",
            "name",
          ],
        },
      ],
      "user": Object {
        "id": "[NUMBER]",
        "name": "Zeb Zob",
        "nodeId": "[NodeId]",
      },
    },
  },
}
`;
